providers/implementations/encode_decode/decode_der2key.c

```diff
@@ -544,15 +544,32 @@ static void *rsa_d2i_PKCS8(void **key, const unsigned char **der, long der_len,
 
 static int rsa_check(void *key, struct der2key_ctx_st *ctx)
 {
+    int	valid;
+    int n, e_ok, d_ok, p_ok, q_ok, dmp1_ok, dmq1_ok, iqmp_ok;
+
     switch (RSA_test_flags(key, RSA_FLAG_TYPE_MASK)) {
     case RSA_FLAG_TYPE_RSA:
-        return ctx->desc->evp_type == EVP_PKEY_RSA;
+        valid = (ctx->desc->evp_type == EVP_PKEY_RSA);
+        break;
     case RSA_FLAG_TYPE_RSASSAPSS:
-        return ctx->desc->evp_type == EVP_PKEY_RSA_PSS;
+        valid = (ctx->desc->evp_type == EVP_PKEY_RSA_PSS);
+        break;
+    default:
+        /* Currently unsupported RSA key type */
+        valid = 0;
     }
 
-    /* Currently unsupported RSA key type */
-    return 0;
+    n = BN_num_bits(RSA_get0_n(key));
+    e_ok = (BN_num_bits(RSA_get0_e(key)) < n);
+    d_ok = (BN_num_bits(RSA_get0_d(key)) < n);
+    p_ok = (BN_num_bits(RSA_get0_p(key)) < n);
+    q_ok = (BN_num_bits(RSA_get0_q(key)) < n);
+    dmp1_ok = (BN_num_bits(RSA_get0_dmp1(key)) < n);
+    dmq1_ok = (BN_num_bits(RSA_get0_dmq1(key)) < n);
+    iqmp_ok = (BN_num_bits(RSA_get0_iqmp(key)) < n);
+    valid = (valid && e_ok && d_ok && p_ok && q_ok && dmp1_ok && dmq1_ok && iqmp_ok);
+
```

Comment:
2 questions

1) Is this sufficient for all RSA key cases?  For instance what if the buffer happens to decode to a multiprime key, do those key sizes need to be checked as well?  Or is that so unlikely as to not worry about it?
 
2) perhaps more important, Is this ok to check in the decoder like this?  I'm more wondering if there is a use case for openssl to decode invalid keys.

For example, if someone produces an RSA  key with an obscenely large public exponent like so:
```
openssl genpkey -algorithm RSA -pkeyopt rsa_keygen_bits:512 -pkeyopt rsa_keygen_pubexp:0xfffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffffff
```
And then a user wants to check the PEM file they received from this user with
```
openssl pkey -in <file> -text -noout
```
Won't this change just result in them getting an error message as e_ok will evaluate to 0 (since the e has more bits than p or q).  Its actually (I think) a valid key to have that, just really silly to use, and users might want to visually inspect oddly formed keys like that, and (again I think), this would prevent such an operation.

Not saying these aren't valid checks to preform for the purpose of the fuzzer, but it might be better implemented inside the fuzzer using EVP_PKEY_to_data, and checking the returned parameters for fuzzer-specific sanity there

Reply:
They shouldn't be, rebase on master and they should go away.  I think my questions regarding where to do these checks are still valid though

Reply:
Besides the glitch in less-than vs. less-than/equal-to there was a problem with NULL pointer dereference which fortunately got discovered by our tests. Looks like not every RSA-key comes with all its parameters set as indicated by callstack below.

the stack comes from keymgmt_internal_test (`egdb --args test/keymgmt_internal_test test/certs/ee-cert.pem`)
```
egdb --args test/keymgmt_internal_test test/certs/ee-cert.pem

Program received signal SIGSEGV, Segmentation fault.
0x0000008f041216c1 in BN_num_bits (a=0x0) at crypto/bn/bn_lib.c:181
181         int i = a->top - 1;
(gdb) bt
#0  0x0000008f041216c1 in BN_num_bits (a=0x0) at crypto/bn/bn_lib.c:181
#1  0x0000008f0446f4d0 in rsa_check (key=0x9124e16380, ctx=0x917283f3c0) at providers/implementations/encode_decode/decode_der2key.c:564
#2  0x0000008f0446d583 in der2key_decode (vctx=0x917283f3c0, cin=0x9172861810, selection=134, data_cb=0x8f041b9200 <decoder_process>, data_cbarg=0x7f050580be58, pw_cb=0x8f04276d20 <ossl_pw_passphrase_callback_dec>, pw_cbarg=0x9172836038) at providers/implementations/encode_decode/decode_der2key.c:286
#3  0x0000008f041b98de in decoder_process (params=0x7f050580bff0, arg=0x7f050580c258) at crypto/encode_decode/decoder_lib.c:1000
#4  0x0000008f04474d5c in spki2typespki_decode (vctx=0x917284fc80, cin=0x9172861510, selection=134, data_cb=0x8f041b9200 <decoder_process>, data_cbarg=0x7f050580c258, pw_cb=0x8f04276d20 <ossl_pw_passphrase_callback_dec>, pw_cbarg=0x9172836038) at providers/implementations/encode_decode/decode_spki2typespki.c:136
#5  0x0000008f041b98de in decoder_process (params=0x0, arg=0x7f050580c368) at crypto/encode_decode/decoder_lib.c:1000
#6  0x0000008f041b8f29 in OSSL_DECODER_from_bio (ctx=0x9172836000, in=0x9172843f80) at crypto/encode_decode/decoder_lib.c:82
#7  0x0000008f041b9b84 in OSSL_DECODER_from_data (ctx=0x9172836000, pdata=0x7f050580c480, pdata_len=0x7f050580c438) at crypto/encode_decode/decoder_lib.c:157
#8  0x0000008f04388242 in x509_pubkey_ex_d2i_ex (pval=0x9124e1bdd0, in=0x7f050580c6b8, len=421, it=0x8f04522320 <X509_PUBKEY_it.local_it>, tag=-1, aclass=0, opt=0 '\000', ctx=0x7f050580cce0, libctx=0x0, propq=0x0) at crypto/x509/x_pubkey.c:217
#9  0x0000008f04108e56 in asn1_item_embed_d2i (pval=0x9124e1bdd0, in=0x7f050580c6b8, len=421, it=0x8f04522320 <X509_PUBKEY_it.local_it>, tag=-1, aclass=0, opt=0 '\000', ctx=0x7f050580cce0, depth=3, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:262
#10 0x0000008f0410adaa in asn1_template_noexp_d2i (val=0x9124e1bdd0, in=0x7f050580c890, len=421, tt=0x8f04522760 <X509_CINF_seq_tt+240>, opt=0 '\000', ctx=0x7f050580cce0, depth=2, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:682
#11 0x0000008f04109cab in asn1_template_ex_d2i (val=0x9124e1bdd0, in=0x7f050580c890, inlen=421, tt=0x8f04522760 <X509_CINF_seq_tt+240>, opt=0 '\000', ctx=0x7f050580cce0, depth=2, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:558
#12 0x0000008f0410963b in asn1_item_embed_d2i (pval=0x7f050580c9c0, in=0x7f050580c9b8, len=421, it=0x8f04522630 <X509_CINF_it.local_it>, tag=16, aclass=0, opt=0 '\000', ctx=0x7f050580cce0, depth=2, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:422
#13 0x0000008f0410adaa in asn1_template_noexp_d2i (val=0x7f050580c9c0, in=0x7f050580cb90, len=800, tt=0x8f04522840 <X509_seq_tt>, opt=0 '\000', ctx=0x7f050580cce0, depth=1, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:682
#14 0x0000008f04109cab in asn1_template_ex_d2i (val=0x9124e1bd80, in=0x7f050580cb90, inlen=800, tt=0x8f04522840 <X509_seq_tt>, opt=0 '\000', ctx=0x7f050580cce0, depth=1, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:558
#15 0x0000008f0410963b in asn1_item_embed_d2i (pval=0x7f050580ccd8, in=0x7f050580ce08, len=800, it=0x8f04522800 <X509_it.local_it>, tag=16, aclass=0, opt=0 '\000', ctx=0x7f050580cce0, depth=1, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:422
#16 0x0000008f04108791 in asn1_item_ex_d2i_intern (pval=0x7f050580ccd8, in=0x7f050580ce08, len=804, it=0x8f04522800 <X509_it.local_it>, tag=-1, aclass=0, opt=0 '\000', ctx=0x7f050580cce0, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:118
#17 0x0000008f04108878 in ASN1_item_d2i_ex (pval=0x7f050580ccd8, in=0x7f050580ce08, len=804, it=0x8f04522800 <X509_it.local_it>, libctx=0x0, propq=0x0) at crypto/asn1/tasn_dec.c:144
#18 0x0000008f04108906 in ASN1_item_d2i (pval=0x0, in=0x7f050580ce08, len=804, it=0x8f04522800 <X509_it.local_it>) at crypto/asn1/tasn_dec.c:154
#19 0x0000008f04388ee6 in d2i_X509 (a=0x0, in=0x7f050580ce08, len=804) at crypto/x509/x_x509.c:138
#20 0x0000008f0429897c in PEM_ASN1_read_bio (d2i=0x8f04388e90 <d2i_X509>, name=0x8f04015cdc "CERTIFICATE", bp=0x9124e3fb80, x=0x0, cb=0x0, u=0x0) at crypto/pem/pem_oth.c:31
#21 0x0000008f0429cf3c in PEM_read_bio_X509 (bp=0x9124e3fb80, x=0x0, cb=0x0, u=0x0) at crypto/pem/pem_x509.c:18
#22 0x0000008f040f4bcc in test_evp_pkey_export_to_provider (n=0) at test/keymgmt_internal_test.c:318
#23 0x0000008f040f6a13 in run_tests (test_prog_name=0x7f050580e1b8 "/home/sashan/work.openssl/QUIC/openssl.addr_validation/test/keymgmt_internal_test") at test/testutil/driver.c:377
#24 0x0000008f040fb201 in main (argc=2, argv=0x7f050580dfe8) at test/testutil/main.c:31
```

Reply:
We died because of NULL pointer dereference because `RSA::d` member was NULL.

Reply:
Ok, what about my concern regarding the ability to interrogate otherwise invalid keys using openssl Pkey?  That's my big concern

Reply:
this was discussed off-line. the result is: it's not worth to complicate logic with a 'force' flag to suppress the check.

Reply:
> ... Looks like not every RSA-key comes with all its parameters set as indicated by callstack below.

Of course.  RSA public keys only have *N* and *e*.